Add auto metric support for Amplitude

mknowlton89 opened the pull request:
### Features and Changes

<!--
  Include a description of your changes.
  If there are any new tools, API's, etc. that devs can leverage, it may be helpful to include usage info.
-->

<!--
  Indicate which issue this will close, if applicable
  For multiple issues, add a new `- Closes` or `- Fixes` line
-->

- Closes **(add link to issue here)**

### Dependencies

<!--
Please include dependencies that must be met before deploying, if applicable. If none, you can write None or delete this section.
 -->

### Testing

<!--
  Please describe how to test these changes.
 -->

### Screenshots

<!--
  For any UI changes, e.g. changes to /front-end or docs components, please include screenshots
-->

bttf reviewed (approved):
Lgtm. Not sure if we want to put it behind a FF for further testing in prod before releasing?

mknowlton89 commented:
@jdorn Feedback from yesterday has been addressed, and I've refactored the `getSchemaFormatConfig` method, along with any method utilizing the return value from it to abstract the logic a bit more. 

I've added 1 comment about a pattern that I'm not in love with - open to any suggestions there.

mknowlton89 commented on packages/back-end/src/integrations/SqlIntegration.ts:
```diff
@@ -95,6 +96,7 @@ export default abstract class SqlIntegration
       "segment",
       "rudderstack",
       "ga4",
+      "amplitude",
```
I don't love this pattern. I've created a new type `AutoMetricSchemas` that is a union type of the event tracker types that support auto metrics, and I'd love to use that here, but no matter what I've tried, Typescript isn't happy.

With the current pattern, I now have to maintain the list of event trackers that support auto metrics in two places (here, and in the `AutoMetricSchemas` type declaration), which just doesn't feel good for long-term maintenance.

I've tried the following:
- Build an array of supported event trackers here, and derive the `AutoMetricSchemas` type from each index of that array.
- I've tried to define the `AutoMetricSchemas` by extending `SchemaFormatConfig` and using the `Omit` utility. (This returns an error in the `getSchemaFormatConfig` since we don't have a default case there.
- I've tried simply typing the `supportedEventTrackers` array as a `AutoMetricSchemas[]` array, but then the check on line 104 isn't happy.

Open to suggestions

jdorn replied:
I would do your last bullet point.  On line 104, you can make it work with an `as`

```ts
const supportedEventTrackers: AutoMetricSchemas[] = [...];

supportedEventTrackers.includes(this.settings.schemaFormat as AutoMetricSchema)
```

jdorn replied:
This doesn't completely solve the problem since you might add a schema to the type definition, but forget to update this array.  One simple option is to make this an object instead of an array.

```ts
const supportEventTrackers: Record<AutoMetricSchema, true> = {
  segment: true,
  ...
}
```

With this, Typescript will throw an error if you forget to update the code here after adding a new schema to the type.

mknowlton89 commented on packages/back-end/types/datasource.d.ts:
```diff
@@ -153,6 +152,8 @@ export type DataSourceSettings = {
   notebookRunQuery?: string;
   informationSchemaId?: string;
+  // Some event trackers (currently only Amplitude) require a projectId to support auto metrics
+  projectId?: string;
```
@jdorn Initially, I did go down the route of adding the `Project Id` to the second page of the `NewDataSourceForm` like we do for Matomo's `siteId`, but in doing so, I saw that there was no way for a user to ever edit this field after they create the data source, so I ultimately opted to add it to the first page.

I've reverted this back to use the `schemaOptions` like we do for Matomo.

From what I can tell, up until now, we haven't ever needed to allow a user to edit these `schemaOptions` because they were only used when initially generating a new data source (to do things like generate the default queries). However, in this case, the `projectId` will be used to generate the default metric query, and auto generated metric queries. So if a user makes a typo, all of their initial metric queries will be incorrect, and they won't be able to use the auto metrics feature at all.

If there is a place to edit this that I just haven't found, let me know. 

Otherwise, if you like the updated pattern, I can explore how we can allow users to edit this info via the `DataSourceForm`.

mknowlton89 replied:
@jdorn I just pushed up [this commit](https://github.com/growthbook/growthbook/pull/1522/commits/b02f907e1c2faefe9a39ec59fa4a7870021d0512) that introduces a new component `EditSchemaOptions` to the `DataSourceForm` component. It's set up similar to how we handle `ConnectionSettings`. Currently, it's only provisioned for `Amplitude`, but can be expanded as necessary.

With this, a user can now edit their Amplitude `projectId` - allowing orgs with existing Amplitude data sources to use auto metrics, and supporting the use case where an org entered the projectId incorrectly, or they need to change the projectId.

mknowlton89 commented on packages/front-end/components/Settings/NewDataSourceForm.tsx:
```diff
@@ -268,7 +268,10 @@ const NewDataSourceForm: FC<{
-      settings,
+      settings: {
+        ...settings,
+        schemaOptions: form.watch("settings.schemaOptions"),
```
With this change, anytime we're working schemaFormat that contains `schemaOptions` the `schemaOptions` will be persisted on the datasource in the DB. 

From what I can tell, we haven't needed to do this previously because the values here were only used on the initial default query generation, and never again, But now, in Amplitude's case, we need to persist the `projectId` for future use in the auto metrics queries.

Open to suggestions on alternatives.

mknowlton89 commented on packages/back-end/src/integrations/SqlIntegration.ts:
```diff
@@ -1467,52 +1468,86 @@ export default abstract class SqlIntegration
+          anonymousIdColumn: "amplitude_id",
+          getMetricTableName: () =>
+            this.generateTablePath(
+              `EVENTS_${this.settings.schemaOptions?.projectId || `*`}`
```
We shouldn't ever actually wildcard this, as we have a check on the frontend that throws an error if the projectId isn't set, but keeping this.

mknowlton89 commented on packages/back-end/src/integrations/SqlIntegration.ts:
```diff
@@ -1590,96 +1630,81 @@ export default abstract class SqlIntegration
+
+    for (const additionalEvent of additionalEvents) {
+      const sql = this.getTrackedEventSql(
+        `'${additionalEvent.eventName}'`,
```
I don't love this - but haven't found a way around it. The backtics are BigQuery specific.

jdorn commented on packages/back-end/src/integrations/SqlIntegration.ts:
```diff
@@ -1467,52 +1468,94 @@ export default abstract class SqlIntegration
-          dateFormat: "yyyy-MM-dd",
+          getMetricTableName: (eventName: string) =>
+            this.generateTablePath(eventName),
+          getDateLimitClause: (start: Date, end: Date) =>
```
This should not return anything.  We should always add a WHERE clause for the `timestampColumn`.  This function should just be for additional filtering, which only GA4 requires

jdorn commented on packages/back-end/src/integrations/SqlIntegration.ts:
```diff
@@ -1590,96 +1630,81 @@ export default abstract class SqlIntegration
+    groupByColumn?: string
+  ) {
+    const end = new Date();
+    const start = subDays(new Date(), 70);
```
Change back to 7
